ts/input/tex/ColumnParser.ts

```diff
@@ -0,0 +1,312 @@
+  ralign: [number, string, string][];   // the row alignment and column width/align when specified
+}
+
+/**
+ * A function to handle a column declaration
+ */
+export type ColumnHandler = (state: ColumnState) => void;
+
+
+/***********************************************************************/
+
+/**
+ * The ColumnParser class for processing array environment column templates.
+ */
+export class ColumnParser {
+
+  /**
+   * The handlers for each column character type (future: can be augmented by \newcolumntype)
+   */
+  public columnHandler: {[c: string]: ColumnHandler} = {
+    l: (state) => state.calign[state.j++] = 'left',
+    c: (state) => state.calign[state.j++] = 'center',
+    r: (state) => state.calign[state.j++] = 'right',
+    p: (state) => this.getColumn(state, TEXCLASS.VTOP),
+    m: (state) => this.getColumn(state, TEXCLASS.VCENTER),
+    b: (state) => this.getColumn(state, TEXCLASS.VBOX),
+    w: (state) => this.getColumn(state, TEXCLASS.VTOP, ''),
+    W: (state) => this.getColumn(state, TEXCLASS.VTOP, ''),
+    '|': (state) => this.addRule(state, 'solid'),
+    ':': (state) => this.addRule(state, 'dashed'),
+    '>': (state) => state.cstart[state.j] = (state.cstart[state.j] || '') + this.getBraces(state),
+    '<': (state) => state.cend[state.j - 1] = (state.cend[state.j - 1] || '') + this.getBraces(state),
+    '@': (state) => this.addAt(state, this.getBraces(state)),
+    '!': (state) => this.addBang(state, this.getBraces(state)),
+    //
+    // Non-standard for math-mode versions
+    //
+    P: (state) => this.macroColumn(state, '>{$}p{#1}<{$}', 1),
+    M: (state) => this.macroColumn(state, '>{$}m{#1}<{$}', 1),
+    B: (state) => this.macroColumn(state, '>{$}b{#1}<{$}', 1),
+    //
+    // Ignored
+    //
+    ' ': (_state) => {},
+  };
+
+  /**
+   * The maximum number of column specifiers to process (prevents loops from \newcolumntype).
+   */
+  public MAXCOLUMNS: number = 10000;
+
+  /**
+   * Process an array column template
+   *
+   * @param {TexParser} parser   The active TexParser
+   * @param {string} template    The alignment template
+   * @param {ArrayItem} array    The ArrayItem for the template
+   */
+  public process(parser: TexParser, template: string, array: ArrayItem) {
+    //
+    // Initialize the state
+    //
+    const state: ColumnState = {
+      parser, template, i: 0, j: 0, c: '',
+      cwidth: [], calign: [], cspace: [], clines: [],
+      cstart: array.cstart, cend: array.cend,
+      ralign: array.ralign, cextra: array.cextra
+    };
+    //
+    // Loop through the template to process the column specifiers
+    //
+    let n = 0;
+    while (state.i < state.template.length) {
+      if (n++ > this.MAXCOLUMNS) {
+        throw new TexError('MaxColumns', 'Too many column specifiers (perhaps looping column definitions?)');
+      }
+      const c = state.c = String.fromCodePoint(state.template.codePointAt(state.i));
```

Comment:
Since `getCodePoint()` works on the parser's `string` property, I can't use it directly, here, but I have included the same check for the undefined `codePointAt()`.  We could add a utility function to `util/strings.ts` that takes a string and an index, and use it both in `getCodePoint()` and here, if you prefer.